staking/programs/publisher-caps/src/lib.rs

```diff
@@ -1,46 +1,121 @@
 
-pub fn get_dummy_publisher(i: usize) -> Pubkey {
-    let mut bytes = [0u8; 32];
-    bytes[0] = (i % 256) as u8;
-    bytes[1] = (i / 256) as u8;
-    Pubkey::from(bytes)
-}
-
 #[program]
 pub mod publisher_caps {
     use super::*;
 
-    // TODO: Actually implement this using wormhole
-    pub fn post_publisher_caps(
-        ctx: Context<PostPublisherCaps>,
-        first_publisher: Pubkey,
-        first_publisher_cap: u64,
-    ) -> Result<()> {
+    pub fn init_publisher_caps(ctx: Context<InitPublisherCaps>) -> Result<()> {
         let publisher_caps = &mut ctx.accounts.publisher_caps.load_init()?;
-        publisher_caps.timestamp = Clock::get()?.unix_timestamp;
-        publisher_caps.num_publishers = MAX_CAPS as u64;
-
-        publisher_caps.caps[0].pubkey = first_publisher;
-        publisher_caps.caps[0].cap = first_publisher_cap;
+        publisher_caps.write_authority = *ctx.accounts.signer.key;
+        publisher_caps.is_verified = 0;
+        Ok(())
+    }
 
-        for i in 1..MAX_CAPS {
-            publisher_caps.caps[i].pubkey = get_dummy_publisher(i);
-            publisher_caps.caps[i].cap = i as u64;
+    pub fn write_publisher_caps(
+        ctx: Context<WritePublisherCaps>,
+        index: u32,
+        data: Vec<u8>,
+    ) -> Result<()> {
+        {
+            let publisher_caps = &mut ctx.accounts.publisher_caps.load()?;
+
+            require_eq!(
+                publisher_caps.is_verified,
+                0,
+                PublisherCapsError::CantMutateVerifiedPublisherCaps
+            );
         }
 
-        // publisher caps should always be sorted
-        publisher_caps.caps.sort();
+        require_gte!(
+            PublisherCaps::LEN,
+            PublisherCaps::HEADER_LEN
+                .saturating_add(index.try_into().unwrap())
+                .saturating_add(data.len()),
+            PublisherCapsError::DataOverflow
+        );
+
+        let account_info = ctx.accounts.publisher_caps.to_account_info();
+        sol_memcpy(
+            &mut account_info.try_borrow_mut_data().unwrap()
+                [PublisherCaps::HEADER_LEN + index as usize..],
+            &data,
+            data.len(),
+        );
+
+        Ok(())
+    }
+
+    pub fn verify_publisher_caps(
+        ctx: Context<VerifyPublisherCaps>,
+        proof: Vec<[u8; 20]>,
+    ) -> Result<()> {
+        let vaa = VaaAccount::load_unchecked(&ctx.accounts.encoded_vaa);
+        let publisher_caps = &mut ctx.accounts.publisher_caps.load_mut()?;
+
+
+        require_eq!(
+            Pubkey::from(vaa.emitter_address()),
+            PRICE_FEEDS_EMITTER_ADDRESS,
+            PublisherCapsError::WrongEmitterAddress
+        );
+        require_eq!(
+            vaa.emitter_chain(),
+            PRICE_FEEDS_EMITTER_CHAIN,
+            PublisherCapsError::WrongEmitterChain
+        );
+
+        publisher_caps.is_verified = 1;
+
+        require_eq!(
+            publisher_caps.discriminator(),
+            2,
```

Comment:
It's better not to have "magic" numbers in the code. Ideally we would want to import this from [where it's defined](https://github.com/pyth-network/pyth-client/pull/412/files#diff-906b53bd04fb93488ed0af575b22bbaec6640813e04dc90b5bedec71f3994776R350).
But let's atleast put it in a const.

Reply:
done